### PR TITLE
Fix #6528: Mirroring Pod conditions to Notebook

### LINKED ISSUE #6528: Notebook conditions are being appended
/kind bug

**What steps did you take and what happened:**
Right now the Notebooks controller will be appending new Pod conditions to the Notebook's status https://github.com/kubeflow/kubeflow/blob/master/components/notebook-controller/controllers/notebook_controller.go#L247. 

This means that a Notebook could end up with the following conditions being present at the same time:
```
        {
          "lastProbeTime": "2022-04-18T08:29:27Z", 
          "message": "Error", 
          "reason": "Error", 
          "type": "Terminated"
        }, 
        {
          "lastProbeTime": "2022-04-15T06:39:50Z", 
          "type": "Running"
        }, 
```

**What did you expect to happen:**
K8s Conditions are used for expressing the state of an object **at a specific point in time**. 
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

We shouldn't be appending the Pod Conditions to the Notebook Conditions, but rather just mirror the Pod conditions to the Notebook.

**Anything else you would like to add:**
If we also add a proper function for converting the PodConditions to Notebook conditions with some basic checks like:
```golang
func PodCondToNotebookCond(c corev1.PodCondition) v1.NotebookCondition {
	notebookCond := v1.NotebookCondition{}

	if len(c.Type) > 0 {
		notebookCond.Type = c.Type
	}

	if len(c.Reason) > 0 {
		notebookCond.Reason = c.Reason
	}
...
```
Then we should be way more solid and also close the relevant, but orthogonal, https://github.com/kubeflow/kubeflow/issues/6056

### PR DESCRIPTION
**Pull request by midhun1998**
Sorry for the delay. I have been bit busy lately. 

I have added the implementation based on @kimwnasptd suggestion to mirror the condition of Notebook Pod and not append container statuses. Correct me if my understanding is incorrect but from what I checked I believe we still have a few fields missing in the Notebook CRD for conditions like "status" and "LastTransitionTime". Let me know if we need to add those too. I can add a PR for that before merging this one in.

### CODE REVIEW

**Review by kimwnasptd — COMMENTED** [review]
Thanks for the PR @midhun1998! I have some nits on renames, but overall LGTM.

I'll also do a test now to verify it works as expected. Also cc @apo-ger since we wanted to include this logic as part of the effort for https://github.com/kubeflow/kubeflow/issues/6056

**Comment by kimwnasptd** [comment]
> Correct me if my understanding is incorrect but from what I checked I believe we still have a few fields missing in the Notebook CRD for conditions like "status" and "LastTransitionTime"

@midhun1998 indeed! We need these two fields to be on par with [Pod Conditions](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#podcondition-v1-core) could you add these two as well as part of your PR?

**Comment by kimwnasptd** [comment]
Then the last item to re-evaluate would be the `status.containerState` of the Notebook CRs. Right now it will always be `{}` 
```yaml
status:
  conditions:
  - lastProbeTime: null
    type: Initialized
  - lastProbeTime: null
    type: Ready
  - lastProbeTime: null
    type: ContainersReady
  - lastProbeTime: null
    type: PodScheduled
  containerState: {}
  readyReplicas: 1
```

But we can tackle this one in the follow-up PR from @apo-ger, for https://github.com/kubeflow/kubeflow/issues/6056, where we are updating how the controller sets the status of the CRs.

File: components/notebook-controller/controllers/notebook_controller.go [hunk 1]
```diff
@@ -241,19 +241,15 @@ func (r *NotebookReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 				if pod.Status.ContainerStatuses[i].State == instance.Status.ContainerState {
 					continue
 				}
-
-				log.Info("Updating Notebook CR state: ", "namespace", instance.Namespace, "name", instance.Name)
-				cs := pod.Status.ContainerStatuses[i].State
-				instance.Status.ContainerState = cs
```
Comment (kimwnasptd) [inline components/notebook-controller/controllers/notebook_controller.go:247]: Lastly, let's keep this logic that updates the `instance.Status.ContainerState` of the CR. The web app is using this information.

**Comment by midhun1998** [comment]
> > Correct me if my understanding is incorrect but from what I checked I believe we still have a few fields missing in the Notebook CRD for conditions like "status" and "LastTransitionTime"
> 
> @midhun1998 indeed! We need these two fields to be on par with [Pod Conditions](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#podcondition-v1-core) could you add these two as well as part of your PR?

Thanks for the confirmation @kimwnasptd ! I was wondering if the changes should go to all 3 APIs (v1, v1alpha1, and v1beta1) or just the v1. Could you please clarify this too?

File: components/notebook-controller/controllers/notebook_controller.go (hunk 1 above)
Comment (midhun1998) [reply components/notebook-controller/controllers/notebook_controller.go:247]: Agree

File: components/notebook-controller/controllers/notebook_controller.go [hunk 2]
```diff
@@ -241,19 +241,15 @@ func (r *NotebookReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
-					oldConditions[0].Reason != newCondition.Reason ||
-					oldConditions[0].Message != newCondition.Message {
-					log.Info("Appending to conditions: ", "namespace", instance.Namespace, "name", instance.Name, "type", newCondition.Type, "reason", newCondition.Reason, "message", newCondition.Message)
-					instance.Status.Conditions = append([]v1beta1.NotebookCondition{newCondition}, oldConditions...)
+				notebookCondition := []v1beta1.NotebookCondition{}
+				for i := range pod.Status.Conditions {
+					log.Info("Updating Notebook CR state: ", "namespace", instance.Namespace, "name", instance.Name)
+					// Mirroring pod condition
+					condition := PodCondToNotebookCond(pod.Status.Conditions[i])
+					log.Info("Mirroring pod conditions: ", "namespace", instance.Namespace, "name", instance.Name, "type", condition.Type, "reason", condition.Reason, "message", condition.Message)
```
[the remark this answers is not in the document]
Comment (midhun1998) [reply components/notebook-controller/controllers/notebook_controller.go:249]: Oops! Will fix this

**Comment by kimwnasptd** [comment]
@midhun1998 I'm just summarizing the changes I proposed, since I did a small back and forth regarding the `containerState`:
1. Let's do the (nit) renamings when logging
2. Let's add the `Status` and `LastTransitionTime` to the API of the notebook conditions. Let's do this for all the versions of the CRD
3. Let's keep the logic that sets the `containerState`, since the web apps use this information

**Comment by midhun1998** [comment]
> @midhun1998 I'm just summarizing the changes I proposed, since I did a small back and forth regarding the `containerState`:
> 
> 1. Let's do the (nit) renamings when logging
> 2. Let's add the `Status` and `LastTransitionTime` to the API of the notebook conditions. Let's do this for all the versions of the CRD
> 3. Let's keep the logic that sets the `containerState`, since the web apps use this information

Thanks for summarizing it @kimwnasptd . The requested changes have been added.

**Comment by kimwnasptd** [comment]
@midhun1998 the changes look good, thank you for your work!

/lgtm
/approve

**Pull request merged**